src/v/coproc/router.cc

```diff
@@ -384,4 +372,109 @@ router::reader_cfg(model::offset start, model::offset end) {
+                    } else {
+                        found->second->scripts.emplace(id);
+                    }
+                    vlog(coproclog.info, "Inserted ntp {} id {}", ntp, id);
+                }
+                src.promise.set_value(errc::success);
+                return ss::now();
+            });
+      });
+}
+
+ss::future<> router::process_removals() {
+    return ss::do_with(
+      std::move(_deferred_removals),
+      [this](absl::flat_hash_map<script_id, ss::promise<bool>>& drm) {
+          return ss::do_for_each(
+            drm.begin(),
+            drm.end(),
+            [this](std::pair<const script_id, ss::promise<bool>>& e) {
+                auto& [id, promise] = e;
+                absl::flat_hash_set<model::ntp> deleted;
+                std::for_each(
+                  _sources.begin(),
+                  _sources.end(),
+                  [&deleted, id = id](
+                    std::pair<const model::ntp, ss::lw_shared_ptr<topic_state>>&
+                      p) {
+                      auto& scripts = p.second->scripts;
+                      scripts.erase(id);
+                      vlog(coproclog.info, "Deleted script id: {}", id);
+                      if (scripts.empty()) {
+                          deleted.emplace(p.first);
+                      }
+                  });
+                // If no more scripts are tracking an ntp, remove the ntp
+                absl::erase_if(
+                  _sources,
+                  [&deleted](const std::pair<
+                             const model::ntp,
+                             ss::lw_shared_ptr<topic_state>>& p) {
+                      return deleted.contains(p.first);
+                  });
+
+                promise.set_value(!deleted.empty());
+            });
+      });
+}
+
+ss::future<> router::cancel_pending_updates() {
+    auto cda = ss::do_with(
+      std::move(_deferred_additions),
+      [](absl::flat_hash_map<script_id, source<errc>>& dfa) {
+          return ss::parallel_for_each(
+            dfa.begin(),
+            dfa.end(),
+            [](std::pair<const script_id, source<errc>>& src) {
+                src.second.promise.set_exception(
+                  std::make_exception_ptr(std::runtime_error(
+                    fmt::format("Pending addition cancelled: {}", src.first))));
+            });
+      });
+    auto cdc = ss::do_with(
+      std::move(_deferred_removals),
+      [](absl::flat_hash_map<script_id, ss::promise<bool>>& drm) {
+          return ss::parallel_for_each(
+            drm.begin(),
+            drm.end(),
+            [](std::pair<const script_id, ss::promise<bool>>& r) {
+                r.second.set_exception(
+                  std::make_exception_ptr(std::runtime_error(
+                    fmt::format("Pending removal cancelled: {}", r.first))));
+            });
+      });
+    return ss::when_all_succeed(std::move(cda), std::move(cdc))
```

Comment:
do we have any benefits from dispatching in the background at the same time? 

also, this doesn't handle _any_ exceptions and it's impossible to get a good trace.